src/main/java/com/zaxxer/nuprocess/osx/ProcessKqueue.java

```diff
@@ -33,13 +36,12 @@
  */
 final class ProcessKqueue extends BaseEventProcessor<OsxProcess>
 {
-   private static final int KEVENT_POOL_SIZE = 16;
+   private static final int NUM_KEVENTS = 64;
    private static final TimeSpec timeSpec;
    private static final int JAVA_PID;
 
    private volatile int kqueue;
-   private Kevent triggeredEvent;
-   private BlockingQueue<Kevent> keventPool;
+   private ThreadLocal<KeventArray> keventArrays;
```

Comment:
Totally agreed, the shut down path was broken before as well. I'll clean it
up.

I'm happy to convert to use a JNA Structure instead of the existing logic
for the kevent size.

I don't feel too strongly about the drainTo() call, but I did check the
source and it is indeed atomic.
2015年9月4日(金) 23:21 Brett Wooldridge <notifications@github.com>:

> In src/main/java/com/zaxxer/nuprocess/osx/ProcessKqueue.java
> <https://github.com/brettwooldridge/NuProcess/pull/31#discussion_r38808240>
> :
>
> >     private static final TimeSpec timeSpec;
> >     private static final int JAVA_PID;
> >
> >     private volatile int kqueue;
> > -   private Kevent triggeredEvent;
> > -   private BlockingQueue<Kevent> keventPool;
> > +   private ThreadLocal<KeventArray> keventArrays;
>
> This not an issue if the only interaction is this ThreadLocal is by a
> NuProcess managed thread (see comments on registerProcess(OsxProcess
> process) below).
>
> However, the BaseEventProcess.run()
> <https://github.com/brettwooldridge/NuProcess/blob/master/src/main/java/com/zaxxer/nuprocess/internal/BaseEventProcessor.java#L61>
> method *must* then have a finally that calls a shutdown/cleanup method so
> that:
>
>    - KeventArray.free() gets called, and
>    - keventArrays.remove() is called to remove the ThreadLocal
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/brettwooldridge/NuProcess/pull/31/files#r38808240>.
>